README.md
# log-viewer

[![Build Status](https://badgen.net/travis/FEMessage/log-viewer/master)](https://travis-ci.com/FEMessage/log-viewer)
[![NPM Download](https://badgen.net/npm/dm/@femessage/log-viewer)](https://www.npmjs.com/package/@femessage/log-viewer)
[![NPM Version](https://badgen.net/npm/v/@femessage/log-viewer)](https://www.npmjs.com/package/@femessage/log-viewer)
[![NPM License](https://badgen.net/npm/license/@femessage/log-viewer)](https://github.com/FEMessage/log-viewer/blob/master/LICENSE)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/FEMessage/log-viewer/pulls)
[![Automated Release Notes by gren](https://img.shields.io/badge/%F0%9F%A4%96-release%20notes-00B2EE.svg)](https://github-tools.github.io/github-release-notes/)

log-viewer is a vue component which can display terminal log in browser with high performance.

![new-log](https://user-images.githubusercontent.com/53422750/65296811-4cb5b700-db98-11e9-9b55-1a5c8633ae8f.gif)

[中文文档](./README-zh.md)

## Table of Contents

- [Features](#features)
- [Install](#install)
- [Usage](#usage)
- [Links](#links)
- [Reference](#reference)
- [Performance](#performance)
  - [Memory Usage](#memory-usage)
    - [item-mode](#item-mode)
    - [vfor-mode](#vfor-mode)
  - [Render timeline](#render-timeline)
    - [item-mode](#item-mode-1)
    - [vfor-mode](#vfor-mode-1)
- [Contributing](#contributing)
- [Contributors](#contributors)
- [License](#license)

## Features

- Process some special characters in the log stream
- High performance, and process large amounts of data without jamming
- Customize loading status
- Auto scroll to the bottom

[⬆ Back to Top](#table-of-contents)

## Install

```bash
yarn add @femessage/log-viewer
```

[⬆ Back to Top](#table-of-contents)

## Usage

```html
<log-viewer :log="log" :loading="isLoading" />
```

[⬆ Back to Top](#table-of-contents)

## Links

- [design doc](https://www.yuque.com/docs/share/db3640ad-ab65-4588-8244-d245f90e9a6a?translate=en)

[⬆ Back to Top](#table-of-contents)

## Reference

- [thanks to react-lazylog](https://github.com/mozilla-frontend-infra/react-lazylog)
- [travis-ci logs](https://travis-ci.org/)
- [http://jafrog.com/2013/11/23/colors-in-terminal.html](http://jafrog.com/2013/11/23/colors-in-terminal.html)
- [https://en.wikipedia.org/wiki/ANSI_escape_code](https://en.wikipedia.org/wiki/ANSI_escape_code)

[⬆ Back to Top](#table-of-contents)

## Performance

Virtual scrolling use the component：[https://github.com/tangbc/vue-virtual-scroll-list](https://github.com/tangbc/vue-virtual-scroll-list)

Achieve internal use 'item-mode' so that its performance is better than the 'vfor-mode'.

### Memory Usage

Use 100,000 lines to test.

#### item-mode

The Memory before the component mount: 36.5MB<br />
The Memory after the component mounted: 48MB<br />
Memory Usage: 11.5MB<br />
![图片.gif](https://cdn.nlark.com/yuque/0/2019/gif/298847/1563439757710-ffb3b170-839c-46ca-810e-06e041fe93bd.gif#align=left&display=inline&height=425&name=%E5%9B%BE%E7%89%87.gif&originHeight=425&originWidth=720&size=266064&status=done&width=720)

#### vfor-mode

The Memory before the component mount: 43MB<br />
The Memory after the component mounted: 221MB<br />
Memory Usage: 178MB

![item-10e4.gif](https://cdn.nlark.com/yuque/0/2019/gif/298847/1563439644472-91058ae0-804e-4d03-bd9c-3cb116283a3c.gif#align=left&display=inline&height=425&name=item-10e4.gif&originHeight=425&originWidth=720&size=270167&status=done&width=720)

### Render timeline

Also use 100,000 lines to test.

#### item-mode

Render time: 0.63ms<br />
Patch time: 72.18ms<br />
Total time: 72.85ms<br />
![item-1e5.gif](https://cdn.nlark.com/yuque/0/2019/gif/298847/1563849440263-9ee2e04b-eac3-49fc-8ec5-b4ac63f77f8b.gif#align=left&display=inline&height=540&name=item-1e5.gif&originHeight=540&originWidth=1081&size=175825&status=done&width=1081)

#### vfor-mode

Render time: 933.05ms<br />
Patch time: 23.81ms<br />
Total time: 956.86ms<br />
![v-for-1e5.gif](https://cdn.nlark.com/yuque/0/2019/gif/298847/1563849485762-ea08a525-f04c-4827-ab1d-5242d92b80ba.gif#align=left&display=inline&height=811&name=v-for-1e5.gif&originHeight=811&originWidth=1211&size=94294&status=done&width=1211)

[⬆ Back to Top](#table-of-contents)

## Contributing

For those who are interested in contributing to this project, such as:

- report a bug
- request new feature
- fix a bug
- implement a new feature

Please refer to our [contributing guide](https://github.com/FEMessage/.github/blob/master/CONTRIBUTING.md).

[⬆ Back to Top](#table-of-contents)

## Contributors

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore -->
<table><tr><td align="center"><a href="https://github.com/snowlocked"><img src="https://avatars0.githubusercontent.com/u/19562649?v=4" width="100px;" alt="snowlocked"/><br /><sub><b>snowlocked</b></sub></a><br /><a href="https://github.com/FEMessage/log-viewer/commits?author=snowlocked" title="Code">💻</a> <a href="https://github.com/FEMessage/log-viewer/commits?author=snowlocked" title="Documentation">📖</a> <a href="https://github.com/FEMessage/log-viewer/commits?author=snowlocked" title="Tests">⚠️</a></td><td align="center"><a href="https://evila.me"><img src="https://avatars3.githubusercontent.com/u/19513289?v=4" width="100px;" alt="EVILLT"/><br /><sub><b>EVILLT</b></sub></a><br /><a href="https://github.com/FEMessage/log-viewer/commits?author=evillt" title="Code">💻</a></td></tr></table>

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

[⬆ Back to Top](#table-of-contents)

## License

[MIT](./LICENSE)

[⬆ Back to Top](#table-of-contents)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.5.0](https://github.com/FEMessage/log-viewer/compare/v1.4.2...v1.5.0) (2020-11-11)


### Features

* add types ([#18](https://github.com/FEMessage/log-viewer/issues/18)) ([9dd82f0](https://github.com/FEMessage/log-viewer/commit/9dd82f0))



### [1.4.2](https://github.com/FEMessage/log-viewer/compare/v1.4.1...v1.4.2) (2020-08-13)



### [1.4.1](https://github.com/FEMessage/log-viewer/compare/v1.4.0...v1.4.1) (2020-02-13)


### Bug Fixes

* safari中，使用white-space时，超出一行的文字重叠了 ([#14](https://github.com/FEMessage/log-viewer/issues/14)) ([ba3289d](https://github.com/FEMessage/log-viewer/commit/ba3289d))



## [1.4.0](https://github.com/FEMessage/log-viewer/compare/v1.3.0...v1.4.0) (2019-12-03)


### Features

* **scroll-duration:** add animate effect to auto scroll ([#11](https://github.com/FEMessage/log-viewer/issues/11)) ([7691b56](https://github.com/FEMessage/log-viewer/commit/7691b56))
* Terminal colors ([#10](https://github.com/FEMessage/log-viewer/issues/10)) ([9f5408f](https://github.com/FEMessage/log-viewer/commit/9f5408f))



## [1.3.0](https://github.com/FEMessage/log-viewer/compare/v1.2.3...v1.3.0) (2019-11-29)


### Bug Fixes

* 开头空格不显示 ([#9](https://github.com/FEMessage/log-viewer/issues/9)) ([9f19faa](https://github.com/FEMessage/log-viewer/commit/9f19faa))


### Features

* Terminal colors ([#10](https://github.com/FEMessage/log-viewer/issues/10)) ([9f5408f](https://github.com/FEMessage/log-viewer/commit/9f5408f))



### [1.2.3](https://github.com/FEMessage/log-viewer/compare/v1.2.2...v1.2.3) (2019-11-22)


### Bug Fixes

* 开头空格不显示 ([#9](https://github.com/FEMessage/log-viewer/issues/9)) ([9f19faa](https://github.com/FEMessage/log-viewer/commit/9f19faa))



### [1.2.2](https://github.com/FEMessage/log-viewer/compare/v1.2.1...v1.2.2) (2019-10-18)


### Bug Fixes

* adjust loading animation ([#7](https://github.com/FEMessage/log-viewer/issues/7)) ([5d59beb](https://github.com/FEMessage/log-viewer/commit/5d59beb))



### [1.2.1](https://github.com/FEMessage/log-viewer/compare/v1.2.0...v1.2.1) (2019-09-20)



## [1.2.0](https://github.com/FEMessage/log-viewer/compare/v1.1.0...v1.2.0) (2019-09-20)


### Features

* 变更 loading 样式为 3 个小圆点 ([#4](https://github.com/FEMessage/log-viewer/issues/4)) ([e96beea](https://github.com/FEMessage/log-viewer/commit/e96beea))



## [1.1.0](https://github.com/FEMessage/log-viewer/compare/v1.0.1...v1.1.0) (2019-09-20)


### Features

* 变更 loading 样式为 3 个小圆点 ([#4](https://github.com/FEMessage/log-viewer/issues/4)) ([e96beea](https://github.com/FEMessage/log-viewer/commit/e96beea))



### 1.0.1 (2019-07-29)


test/utils/index.test.js
/**
 * 日志输出测试
 * 参考字符串来源：https://api.travis-ci.com/v3/job/196515104/log.txt
 */

import {split2Lines} from '@/utils/index.js'

const getStrCharCodes = str => {
  let codes = []
  for (let i = 0; i < str.length; i++) {
    codes.push(str.charCodeAt(i))
  }
  return codes
}

describe('src/utils/index.js', () => {
  it('根据换行符切割字符串', () => {
    const str =
      'It should be split two lines.\n' + 'It should be split two lines.'
    const result = split2Lines(str)

    expect(result[0]).toBe('It should be split two lines.')
    expect(result[1]).toBe('It should be split two lines.')
  })

  it('当Backspace出现在换行符后不会切割字符串', () => {
    const str = 'It should be\n\u0008 only one line.'
    const result = split2Lines(str)
    let resultCharCode = getStrCharCodes(result[0])
    expect(result.length).toBe(1)
    expect([...resultCharCode]).toStrictEqual(getStrCharCodes(str))
  })
})


test/utils/ansi-parse.test.js
import ansiParse from '@/utils/ansi-parse.js'

describe('src/utils/ansi-parse.js', () => {
  it('解释日志字符串为可读数据', () => {
    const log =
      'normal\x1b[30mcolor is black\x1b[0m\x1b[31;1mcolor is red and bold is true\x1b[0m\x1b[32;3mcolor is green and italic is true\x1b[0m\x1b[33;4mcolor is yellow and underline is true\x1b[0m\x1b[30;47mcolor is black and background is white\x1b[0m\x1b[30;39mcolor is reset\x1b[0m\x1b[40;49mbackground is reset\x1b[0m\x1b[1;22mbold is false\x1b[0m\x1b[3;23mitalic is false\x1b[0m\x1b[4;24munderline is false\x1b[0m'
    const result = ansiParse(log)

    expect(result[0]).toStrictEqual({
      text: 'normal'
    })
    expect(result[1]).toStrictEqual({
      foreground: 'black',
      text: 'color is black'
    })
    expect(result[2]).toStrictEqual({
      foreground: 'red',
      bold: true,
      text: 'color is red and bold is true'
    })
    expect(result[3]).toStrictEqual({
      foreground: 'green',
      italic: true,
      text: 'color is green and italic is true'
    })
    expect(result[4]).toStrictEqual({
      foreground: 'yellow',
      underline: true,
      text: 'color is yellow and underline is true'
    })
    expect(result[5]).toStrictEqual({
      foreground: 'black',
      background: 'white',
      text: 'color is black and background is white'
    })
    expect(result[6]).toStrictEqual({
      text: 'color is reset'
    })
    expect(result[7]).toStrictEqual({
      text: 'background is reset'
    })
    expect(result[8]).toStrictEqual({
      bold: false,
      text: 'bold is false'
    })
    expect(result[9]).toStrictEqual({
      italic: false,
      text: 'italic is false'
    })
    expect(result[10]).toStrictEqual({
      underline: false,
      text: 'underline is false'
    })
  })
})


notify.sh
#!/bin/sh
if [ "$TRAVIS_TEST_RESULT" != "0" ]
then
  echo "build not success, bye"
  exit 1
fi

git remote add github https://$GITHUB_TOKEN@github.com/FEMessage/log-viewer.git > /dev/null 2>&1
git push github HEAD:master --follow-tags

GREN_GITHUB_TOKEN=$GITHUB_TOKEN yarn release

url=https://api.github.com/repos/FEMessage/log-viewer/releases/latest
resp_tmp_file=resp.tmp

curl -H "Authorization: token $GITHUB_TOKEN" $url > $resp_tmp_file

html_url=$(sed -n 5p $resp_tmp_file | sed 's/\"html_url\"://g' | awk -F '"' '{print $2}')
body=$(grep body < $resp_tmp_file | sed 's/\"body\"://g;s/\"//g')
version=$(echo $html_url | awk -F '/' '{print $NF}')

msg='{"msgtype": "markdown", "markdown": {"title": "log-viewer更新", "text": "@所有人\n# [log-viewer('$version')]('$html_url')\n'$body'"}}'

curl -X POST https://oapi.dingtalk.com/robot/send\?access_token\=$DINGTALK_ROBOT_TOKEN -H 'Content-Type: application/json' -d "$msg"

rm $resp_tmp_file


src/index.js
// Import vue component
import Component from './log-viewer.vue'

// `Vue.use` automatically prevents you from using
// the same plugin more than once,
// so calling it multiple times on the same plugin
// will install the plugin only once
Component.install = Vue => {
  Vue.component(Component.name, Component)
}

// To auto-install when vue is found
let GlobalVue = null
if (typeof window !== 'undefined') {
  GlobalVue = window.Vue
} else if (typeof global !== 'undefined') {
  GlobalVue = global.Vue
}
if (GlobalVue) {
  GlobalVue.use(Component)
}

// To allow use as module (npm/webpack/etc.) export component
export default Component

// It's possible to expose named exports when writing components that can
// also be used as directives, etc. - eg. import { RollupDemoDirective } from 'rollup-demo';
// export const RollupDemoDirective = component;


src/log-viewer.d.ts
import Vue, {VueConstructor} from 'vue'

declare module '@femessage/log-viewer' {
  class FemessageComponent extends Vue {
    static install(vue: typeof Vue): void
  }

  type CombinedVueInstance<
    Instance extends Vue,
    Data,
    Methods,
    Computed,
    Props
  > = Data & Methods & Computed & Props & Instance

  type ExtendedVue<
    Instance extends Vue,
    Data,
    Methods,
    Computed,
    Props
  > = VueConstructor<
    CombinedVueInstance<Instance, Data, Methods, Computed, Props> & Vue
  >

  type Combined<Data, Methods, Computed, Props> = Data &
    Methods &
    Computed &
    Props

  type LogViewerData = {
    start: number
    scrollStart: number
    animate: any
    LineWrapper: any
  }

  type LogViewerMethods = {}

  type LogViewerComputed = {}

  type LogViewerProps = {
    virtualAttrs: object
    rowHeight: number
    height: number
    log: string
    loading: boolean
    autoScroll: boolean
    hasNumber: boolean
    scrollDuration: number
  }

  type LogViewer = Combined<
    LogViewerData,
    LogViewerMethods,
    LogViewerComputed,
    LogViewerProps
  >

  export interface LogViewerType extends FemessageComponent, LogViewer {}

  const LogViewerConstruction: ExtendedVue<
    Vue,
    LogViewerData,
    LogViewerMethods,
    LogViewerComputed,
    LogViewerProps
  >

  export default LogViewerConstruction
}


styleguide.config.js
const {VueLoaderPlugin} = require('vue-loader')
const path = require('path')
const glob = require('glob')

const sections = (() => {
  const docs = glob
    .sync('docs/*.md')
    .map(p => ({name: path.basename(p, '.md'), content: p}))
  const demos = []
  let faq = '' // 约定至多只有一个faq.md
  const guides = []
  docs.forEach(d => {
    if (/^faq$/i.test(d.name)) {
      d.name = d.name.toUpperCase()
      faq = d
    } else if (/^guide-/.test(d.name)) {
      guides.push(d)
    } else {
      demos.push(d)
    }
  })
  return [
    {
      name: 'Components',
      components: 'src/*.vue',
      usageMode: 'expand'
    },
    {
      name: 'Demo',
      sections: demos,
      sectionDepth: 2
    },
    ...(faq ? [faq] : []),
    ...(guides.length ? [{name: 'Guide', sections: guides}] : [])
  ]
})()

module.exports = {
  styleguideDir: 'docs',
  pagePerSection: true,
  ribbon: {
    url: 'https://github.com/FEMessage/log-viewer'
  },
  sections,
  webpackConfig: {
    module: {
      rules: [
        {
          test: /\.vue$/,
          loader: 'vue-loader'
        },
        {
          test: /\.js?$/,
          exclude: /node_modules/,
          loader: 'babel-loader'
        },
        {
          test: /\.css$/,
          loaders: ['style-loader', 'css-loader']
        },
        {
          test: /\.less$/,
          loaders: ['vue-style-loader', 'css-loader', 'less-loader']
        },
        {
          test: /\.(woff2?|eot|[ot]tf)(\?.*)?$/,
          loader: 'file-loader'
        }
      ]
    },
    plugins: [new VueLoaderPlugin()]
  }
}


src/utils/ansi-parse.js
/* eslint-disable no-plusplus, no-continue */
// CSI解释器，参考
// 字体颜色https://github.com/mozilla-frontend-infra/react-lazylog/blob/master/src/ansiparse.js
const foregroundColors = {
  '30': 'black',
  '31': 'red',
  '32': 'green',
  '33': 'yellow',
  '34': 'blue',
  '35': 'magenta',
  '36': 'cyan',
  '37': 'white',
  '90': 'bright-black',
  '91': 'bright-red',
  '92': 'bright-green',
  '93': 'bright-yellow',
  '94': 'bright-blue',
  '95': 'bright-magenta',
  '96': 'bright-cyan',
  '97': 'bright-white'
}
// 字体背景
const backgroundColors = {
  '40': 'black',
  '41': 'red',
  '42': 'green',
  '43': 'yellow',
  '44': 'blue',
  '45': 'magenta',
  '46': 'cyan',
  '47': 'white',
  '100': 'bright-black',
  '101': 'bright-red',
  '102': 'bright-green',
  '103': 'bright-yellow',
  '104': 'bright-blue',
  '105': 'bright-magenta',
  '106': 'bright-cyan',
  '107': 'bright-white'
}
// 文字状态（粗体，斜体，下划线）
const styles = {
  '1': 'bold',
  '3': 'italic',
  '4': 'underline'
}
// Select Graphic Rendition(SGR) flag
const SGRCode = 'm'
// 非SGR，后续可针对该控制符进行解释
const notSGRCodes = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'J',
  'K',
  'S',
  'T',
  'f',
  's',
  'u',
  'h',
  'l'
]

/**
 * 格式化数据类型
 * interface formatter {
 *   text: string,
 *   foreground?: string,
 *   background?: string,
 *   underline?: boolean,
 *   bold?: boolean,
 *   italic?: boolean,
 * }
 **/

/**
 * Backspace操作,即删除最后一个字符
 * matchingText: string,
 * result: formatter[]
 * eraseChar: [string,formatter[]]
 * */

const eraseChar = (matchingText, result) => {
  if (matchingText.length) {
    return [matchingText.substr(0, matchingText.length - 1), result]
  } else if (result.length) {
    const index = result.length - 1
    const {text} = result[index]
    const newResult =
      text.length === 1
        ? result.slice(0, result.length - 1)
        : result.map((item, i) =>
            index === i
              ? {...item, text: text.substr(0, text.length - 1)}
              : item
          )

    return [matchingText, newResult]
  }

  return [matchingText, result]
}

/**
 * str: string
 * ansiParse: formatter[]
 * This Function only translate these codes: 1,3,4,22-24,30-37,39,40-47,49,90-97,100-107.
 * It would be more and more powerful and translate more codes.
 **/

const ansiParse = str => {
  let matchingControl = null
  let matchingCode = null
  let matchingText = ''
  let ansiState = []
  let result = []
  let state = {}

  for (let i = 0; i < str.length; i++) {
    if (matchingControl !== null) {
      if (matchingControl === '\x1b' && str[i] === '[') {
        if (matchingText) {
          state.text = matchingText
          result.push(state)
          state = {}
          matchingText = ''
        }

        matchingControl = null
        matchingCode = ''
      } else {
        matchingText += matchingControl + str[i]
        matchingControl = null
      }

      continue
    } else if (matchingCode !== null) {
      if (str[i] === ';') {
        ansiState.push(matchingCode)
        matchingCode = ''
      } else if (str[i] === SGRCode) {
        ansiState.push(matchingCode)
        matchingCode = null
        matchingText = ''

        for (let a = 0; a < ansiState.length; a++) {
          const ansiCode = +ansiState[a]
          if (foregroundColors[ansiCode]) {
            state.foreground = foregroundColors[ansiCode]
          } else if (backgroundColors[ansiCode]) {
            state.background = backgroundColors[ansiCode]
          } else if (ansiCode === 39) {
            delete state.foreground
          } else if (ansiCode === 49) {
            delete state.background
          } else if (styles[ansiCode]) {
            state[styles[ansiCode]] = true
          } else if (ansiCode === 22) {
            state.bold = false
          } else if (ansiCode === 23) {
            state.italic = false
          } else if (ansiCode === 24) {
            state.underline = false
          }
        }

        ansiState = []
      } else if (notSGRCodes.indexOf(str[i]) > -1) {
        // Ignore codes which is not SGR code and delete them.
        // It should be translated in some day.
        matchingCode = ''
        ansiState = []
      } else {
        matchingCode += str[i]
      }

      continue
    }

    if (str[i] === '\x1b') {
      // ESC Control
      matchingControl = str[i]
    } else if (str[i] === '\u0008') {
      // Backspace Control
      ;[matchingText, result] = eraseChar(matchingText, result)
    } else {
      matchingText += str[i]
    }
  }

  if (matchingText) {
    state.text = matchingText + (matchingControl || '')
    result.push(state)
  }

  return result
}

export default ansiParse
